webapp/src/ts/services/uhc-stats.service.ts

```diff
@@ -0,0 +1,115 @@
+import { Injectable } from '@angular/core';
+import * as moment from 'moment';
+import { isObject as _isObject, uniq as _uniq } from 'lodash-es';
+import * as CalendarInterval from '@medic/calendar-interval';
+
+import { DbService } from '@mm-services/db.service';
+import { ContactTypesService } from '@mm-services/contact-types.service';
+import { SessionService } from '@mm-services/session.service';
+import { AuthService } from '@mm-services/auth.service';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class UHCStatsService {
+  private readonly permission = 'can_view_uhc_stats';
+  private canViewUHCStats;
+
+  constructor(
+    private dbService: DbService,
+    private contactTypesService: ContactTypesService,
+    private sessionService: SessionService,
+    private authService: AuthService
+  ) { }
+
+  private async getLastVisitedDate(contactId) {
+    const records = await this.dbService
+      .get()
+      .query('medic-client/contacts_by_last_visited', { reduce: true, group: true, key: contactId });
+
+    const lastVisitedDateRow = records?.rows?.length ? records.rows[0] : {};
+
+    return _isObject(lastVisitedDateRow.value) ? lastVisitedDateRow.value.max : lastVisitedDateRow.value;
+  }
+
+  private async getVisitsInDateRange(dateRange: DateRange, contactId) {
+    const records = await this.dbService
+      .get()
+      .query(
+        'medic-client/visits_by_date',
+        { start_key: [ contactId, dateRange.start ], end_key: [ contactId, dateRange.end ] }
+      );
+
+    const visits = records?.rows?.map(row => {
+      return moment(row.key[1])
+        .startOf('day')
+        .valueOf();
+    });
+
+    return _uniq(visits);
+  }
+
+  private async canUserViewUHCStats() {
+    if (this.canViewUHCStats !== undefined) {
+      return this.canViewUHCStats;
+    }
+
+    // Disable UHC for DB admins.
+    this.canViewUHCStats = this.sessionService.isDbAdmin() ? false : await this.authService.has(this.permission);
+
+    return this.canViewUHCStats;
+  }
+
+  getUHCInterval(visitCountSettings: VisitCountSettings): DateRange {
+    if (!visitCountSettings) {
+      return;
+    }
+
+    return CalendarInterval.getCurrent(visitCountSettings.monthStartDate);
+  }
+
+  async getHomeVisitStats(contact, visitCountSettings: VisitCountSettings): Promise<VisitStats> {
+    if (!visitCountSettings || !contact) {
+      return;
```

Comment:
I'm wondering how come TS is not complaining about this. The return type is `Promise<VisitStats>` but  here you're returning `undefined` (and no promise).

According to the TS playground, this should have failed: https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgGrAM7DAZTHMDZAbwChlkAbODMdLSAEwBECIAuZEAVwFsAjaAG5kAelHIAKsF4RacXgAdyyBAHtu4TjwHCV6zWADiauJW19BUIaQC+pUupC1Va52ABiIZAF5kNAE8QBGQACkU4KAV2HSsASnYABSg1XkwIAB56bDwCDAA+X0KyCmAYMIBCCKjeOJIVCigIMG4oEBsKexUmlrb6igpqWmymVkhOAEYABgAaBtdDTgAmOYGF8BMzTgBmVeR7WxtSQODkGE0EMGA3M5AYEHDI6NjoBOTU9KzMHPxCYpUypVqgo6iUBj1Wu0VF1Gs1If0BkM6N9RmxJrN5gYtMgVpiNBtTOZkLtoXYgA

Reply:
Found it!
It's because we don't have the project configured with ["strictNullChecks": true,](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#--strictnullchecks)
I've added it to test and got the build broken with tons of typing errors including the one that you highlighted in this service 😅

Build:
<img width="876" alt="Screen Shot 2021-04-16 at 4 26 00 pm" src="https://user-images.githubusercontent.com/66472237/115004500-10138400-9ed1-11eb-90ea-910d2f8b5175.png">

IDE:
<img width="722" alt="Screen Shot 2021-04-16 at 4 26 14 pm" src="https://user-images.githubusercontent.com/66472237/115004429-fe31e100-9ed0-11eb-834b-dcfcf48224bb.png">

Reply:
Added return type 'undefined' to the service's public functions :)